coding-the-legislation/35_periods.md

```diff
@@ -41,58 +118,14 @@ You can generate any period with the following properties and methods:
 
-## Calculating dependencies for a period different than the one they are defined for
-
-Variables and formulas are usually defined for months or years. The first line of a formula usualy precises what kind of period it is working with:
-
-```py
-    # Formula defined for years
-    def function(person, period):
-        period = period.this_year
-        ...
-
-    # Formula defined for months
-    def function(person, period):
-        period = period.this_year
-        ...
-
-    # Formula that can handle both months and years
-    def function(person, period):
-        [Ø]
-        ...
-```
-
-Calling a formula with a period it is no able to handle will cause an error. For instance, if we assume that a person `salary` is paid monthly:
-
-```py
-def function(person, period):
-    salary_past_year = person('salary', period.last_year) # THIS WILL BREAK !
-    # >>> Requested period (...) differs from (...) returned by variable salary
-```
-
-However, sometimes, we do need to estimate a variable for a different period that the one it is defined for. 
 
-We may for example want to get the sum of the salaries perceived on the past year, or the past 3 months. The `ADD` option allows you to do it:
+## Automatically set input of compute variables on periods conflicting with their attribute `definition_period`
 
-```py
-def function(person, period):
-    period = period.this_month
-    salary_last_year = person('salary', period.last_year, options = [ADD])
-    salary_last_3_months = person('salary', period.last_3_months, options = [ADD])
-```
+Is is possible to set a montly input variable for a year using the class attribute `set_input`:
 
-The `DIVIDE` option allows you to do the opposite: evaluating a quantity for a smaller period than what it is defined for. For instance, in the following example, `yearly_tax_projected` will contain the value of `some_yearly_tax` for the year including `period` divided by 12.
+* if `set_input = set_input_divide_by_period`, the 12 months are set equal to the 12th of the input value,
+* if `set_input = set_input_dispatch_by_period`, the 12 months are set equal to input value.
 
-```py
-def function(person, period):
-    period = period.this_month
-    tax_projected = person('some_yearly_tax', period.this_month, options = [DIVIDE])
-```
+Is is possible to compute a monthly output variable on a year using the class attribute `calculate_output`:
 
-`ADD` and `DIVIDE` can be both used together in tricky cases, for instance if you want an estimation for the past three month of a variable that is defined yearly :
-
-```py
-def function(person, period):
-    period = period.this_month
-    tax_projected_on_last_3_months = person('some_yearly_tax', period.last_3_months, options = [ADD, DIVIDE])
-```
+* if `calculate_output = calculate_output_add`, the variable is computed for the 12 months and the results are added up. The option ADD is not needed.
```

Comment:
This is misleading. Even if I set the `calculate_output_add ` attribute of a variable, I still need to use the `ADD` option when I calculate it for a too long period, _except_ if I use `calculate_output`, which I can't even do inside a formula with the new syntax. I'm in favor of removing this block, and replacing the whole part by:

--------------------

## Automatically process variable inputs defined for periods not matching the `definition_period`

By default, when you provide a simulation inputs, you won't a able to set a variable value for a period that doesn't match its `definition_period`.

For instance, as the `definition_period` of `salary` is `MONTH`, if you provide as an input a value of `salary` for `2015`, an error will be raised.

It is however possible to define an automatic behaviour to cast yearly inputs into monthy values. To do this, add a `set_input` class attribute to a variable.

* if `set_input = set_input_divide_by_period`, the 12 months are set equal to the 12th of the input value,
* if `set_input = set_input_dispatch_by_period`, the 12 months are set equal to input value.

For instance, let's slightly modify the code of `salary`:
```py
class salary(Variable):
    column = FloatCol
    entity = Person
    label = u"Salary for a month"
    definition_period = MONTH
    set_input = set_input_divide_by_period

    def function(person, period):
        ...
```

We can now provide an input for `2015` for `salary`: no error will be raised, and the value will be automatically split between the 12 months of `2015.